File: agenkit-ts/src/infrastructure/retry-enhanced.ts
/**
 * Enhanced retry logic with:
 * - Multiple jitter types (Full, Equal, Decorrelated)
 * - Per-error-type retry strategies
 * - Budget awareness (cost and count limits)
 * - Backpressure detection
 * - Detailed metrics
 */

import type { Agent, Message } from '../core';

/**
 * Jitter types for retry backoff.
 */
export enum JitterType {
  NONE = 'none',
  FULL = 'full',
  EQUAL = 'equal',
  DECORRELATED = 'decorrelated',
}

/**
 * Error classification for retry strategies.
 */
export enum ErrorClass {
  TRANSIENT = 'transient',
  RATE_LIMIT = 'rate_limit',
  TIMEOUT = 'timeout',
  SERVER_ERROR = 'server_error',
  CLIENT_ERROR = 'client_error',
  UNKNOWN = 'unknown',
}

/**
 * Retry strategy for specific error class.
 */
export interface ErrorStrategy {
  errorClass: ErrorClass;
  maxAttempts: number;
  initialBackoff: number; // milliseconds
  maxBackoff: number; // milliseconds
  backoffMultiplier: number;
  shouldRetry: boolean;
}

/**
 * Retry budget to limit costs.
 */
export interface RetryBudget {
  maxCost: number;
  currentCost: number;
  maxRetriesPerHour: number;
  retryCount: number;
  windowStart: Date;
}

/**
 * Enhanced retry configuration.
 */
export interface EnhancedRetryConfig {
  // Basic retry settings
  maxAttempts: number;
  initialBackoff: number; // milliseconds
  maxBackoff: number; // milliseconds
  backoffMultiplier: number;

  // Jitter settings
  jitterType: JitterType;
  jitterMinRatio: number; // For EqualJitter

  // Error-specific strategies
  errorStrategies: Record<ErrorClass, ErrorStrategy>;
  errorClassifier?: (error: Error) => ErrorClass;

  // Budget settings
  enableBudget: boolean;
  costTracker?: (message: Message) => number;
  maxCostPerHour: number;
  maxRetriesPerHour: number;

  // Backpressure detection
  enableBackpressure: boolean;
  backpressureThreshold: number;
  backpressureWindow: number;
}

/**
 * Default enhanced retry configuration with error strategies.
 */
export function defaultEnhancedRetryConfig(): EnhancedRetryConfig {
  return {
    maxAttempts: 3,
    initialBackoff: 1000,
    maxBackoff: 30000,
    backoffMultiplier: 2.0,
    jitterType: JitterType.FULL,
    jitterMinRatio: 0.5,
    enableBudget: false,
    maxCostPerHour: 100.0,
    maxRetriesPerHour: 1000,
    enableBackpressure: true,
    backpressureThreshold: 0.5,
    backpressureWindow: 100,
    errorStrategies: {
      [ErrorClass.TRANSIENT]: {
        errorClass: ErrorClass.TRANSIENT,
        maxAttempts: 5,
        initialBackoff: 100,
        maxBackoff: 5000,
        backoffMultiplier: 2.0,
        shouldRetry: true,
      },
      [ErrorClass.RATE_LIMIT]: {
        errorClass: ErrorClass.RATE_LIMIT,
        maxAttempts: 10,
        initialBackoff: 60000,
        maxBackoff: 300000,
        backoffMultiplier: 1.5,
        shouldRetry: true,
      },
      [ErrorClass.TIMEOUT]: {
        errorClass: ErrorClass.TIMEOUT,
        maxAttempts: 3,
        initialBackoff: 2000,
        maxBackoff: 30000,
        backoffMultiplier: 2.0,
        shouldRetry: true,
      },
      [ErrorClass.SERVER_ERROR]: {
        errorClass: ErrorClass.SERVER_ERROR,
        maxAttempts: 3,
        initialBackoff: 5000,
        maxBackoff: 60000,
        backoffMultiplier: 2.0,
        shouldRetry: true,
      },
      [ErrorClass.CLIENT_ERROR]: {
        errorClass: ErrorClass.CLIENT_ERROR,
        maxAttempts: 1,
        initialBackoff: 0,
        maxBackoff: 0,
        backoffMultiplier: 1.0,
        shouldRetry: false,
      },
      [ErrorClass.UNKNOWN]: {
        errorClass: ErrorClass.UNKNOWN,
        maxAttempts: 3,
        initialBackoff: 1000,
        maxBackoff: 30000,
        backoffMultiplier: 2.0,
        shouldRetry: true,
      },
    },
  };
}

/**
 * Enhanced retry metrics.
 */
export interface EnhancedRetryMetrics {
  totalAttempts: number;
  successfulFirstAttempt: number;
  successfulOnRetry: number;
  failedAfterRetries: number;
  totalRetries: number;
  totalJitterAdded: number; // seconds
  budgetExceededCount: number;
  backpressureDetected: number;
  errorClassCounts: Record<ErrorClass, number>;
  recentResults: boolean[];
}

/**
 * Enhanced retry decorator wraps an agent with enhanced retry logic.
 */
export class EnhancedRetryDecorator implements Agent {
  private agent: Agent;
  private config: EnhancedRetryConfig;
  private metrics: EnhancedRetryMetrics;
  private budget: RetryBudget;

  constructor(agent: Agent, config?: EnhancedRetryConfig) {
    this.agent = agent;
    this.config = config || defaultEnhancedRetryConfig();
    this.metrics = {
      totalAttempts: 0,
      successfulFirstAttempt: 0,
      successfulOnRetry: 0,
      failedAfterRetries: 0,
      totalRetries: 0,
      totalJitterAdded: 0,
      budgetExceededCount: 0,
      backpressureDetected: 0,
      errorClassCounts: {} as Record<ErrorClass, number>,
      recentResults: [],
    };
    this.budget = {
      maxCost: this.config.maxCostPerHour,
      currentCost: 0,
      maxRetriesPerHour: this.config.maxRetriesPerHour,
      retryCount: 0,
      windowStart: new Date(),
    };
  }

  name(): string {
    return this.agent.name();
  }

  capabilities(): string[] {
    return this.agent.capabilities();
  }

  private classifyError(error: Error): ErrorClass {
    if (this.config.errorClassifier) {
      return this.config.errorClassifier(error);
    }

    // Default classification
    const errStr = error.message.toLowerCase();

    if (errStr.includes('rate limit') || errStr.includes('429')) {
      return ErrorClass.RATE_LIMIT;
    } else if (errStr.includes('timeout') || errStr.includes('timed out')) {
      return ErrorClass.TIMEOUT;
    } else if (errStr.includes('500') || errStr.includes('502') || errStr.includes('503')) {
      return ErrorClass.SERVER_ERROR;
    } else if (
      errStr.includes('400') ||
      errStr.includes('401') ||
      errStr.includes('403') ||
      errStr.includes('404')
    ) {
      return ErrorClass.CLIENT_ERROR;
    }

    return ErrorClass.UNKNOWN;
  }

  private getStrategy(errorClass: ErrorClass): ErrorStrategy {
    return (
      this.config.errorStrategies[errorClass] || {
        errorClass,
        maxAttempts: this.config.maxAttempts,
        initialBackoff: this.config.initialBackoff,
        maxBackoff: this.config.maxBackoff,
        backoffMultiplier: this.config.backoffMultiplier,
        shouldRetry: true,
      }
    );
  }

  private calculateBackoff(baseBackoff: number, attempt: number): number {
    let jittered = baseBackoff;

    switch (this.config.jitterType) {
      case JitterType.NONE:
        break;

      case JitterType.FULL:
        jittered = Math.random() * baseBackoff;
        this.metrics.totalJitterAdded += (baseBackoff - jittered) / 1000;
        break;

      case JitterType.EQUAL: {
        const minBackoff = baseBackoff * this.config.jitterMinRatio;
        jittered = minBackoff + Math.random() * (baseBackoff - minBackoff);
        this.metrics.totalJitterAdded += (baseBackoff - jittered) / 1000;
        break;
      }

      case JitterType.DECORRELATED: {
        if (attempt === 1) {
          jittered = baseBackoff;
        } else {
          const previous = this.calculateBackoff(baseBackoff, attempt - 1);
          jittered = Math.random() * previous * 3 + baseBackoff;
          if (jittered > this.config.maxBackoff) {
            jittered = this.config.maxBackoff;
          }
        }
        break;
      }
    }

    return jittered;
  }

  private checkBudget(cost: number): boolean {
    if (!this.config.enableBudget) {
      return true;
    }

    // Reset window if hour has passed
    const hourInMs = 3600000;
    if (Date.now() - this.budget.windowStart.getTime() > hourInMs) {
      this.budget.currentCost = 0;
      this.budget.retryCount = 0;
      this.budget.windowStart = new Date();
    }

    // Check cost budget
    if (this.budget.currentCost + cost > this.budget.maxCost) {
      this.metrics.budgetExceededCount++;
      return false;
    }

    // Check retry count budget
    if (this.budget.retryCount >= this.budget.maxRetriesPerHour) {
      this.metrics.budgetExceededCount++;
      return false;
    }

    return true;
  }

  private checkBackpressure(): boolean {
    if (!this.config.enableBackpressure) {
      return false;
    }

    const recent = this.metrics.recentResults;
    if (recent.length < this.config.backpressureWindow) {
      return false;
    }

    // Calculate failure rate
    const failures = recent.filter((success) => !success).length;
    const failureRate = failures / recent.length;

    if (failureRate > this.config.backpressureThreshold) {
      this.metrics.backpressureDetected++;
      return true;
    }

    return false;
  }

  /**
   * Process a message with enhanced retry logic.
   */
  async process(message: Message): Promise<Message> {
    let lastError: Error | null = null;
    let errorClass: ErrorClass = ErrorClass.UNKNOWN;
    let strategy: ErrorStrategy = this.getStrategy(errorClass);

    for (let attempt = 1; attempt <= this.config.maxAttempts; attempt++) {
      this.metrics.totalAttempts++;

      // Check budget before attempt
      if (this.config.enableBudget && this.config.costTracker) {
        const estimatedCost = this.config.costTracker(message);
        if (!this.checkBudget(estimatedCost)) {
          throw new Error('Retry budget exceeded');
        }
      }

      // Check backpressure
      if (this.checkBackpressure()) {
        await new Promise((resolve) => setTimeout(resolve, 5000));
      }

      try {
        // Process message
        const response = await this.agent.process(message);

        // Success
        if (attempt === 1) {
          this.metrics.successfulFirstAttempt++;
        } else {
          this.metrics.successfulOnRetry++;
        }
        this.metrics.recentResults.push(true);
        if (this.metrics.recentResults.length > this.config.backpressureWindow) {
          this.metrics.recentResults.shift();
        }

        // Track cost
        if (this.config.enableBudget && this.config.costTracker) {
          const cost = this.config.costTracker(message);
          this.budget.currentCost += cost;
        }

        return response;
      } catch (error) {
        // Failure
        lastError = error as Error;

        // Track failure for backpressure
        this.metrics.recentResults.push(false);
        if (this.metrics.recentResults.length > this.config.backpressureWindow) {
          this.metrics.recentResults.shift();
        }

        // Classify error
        errorClass = this.classifyError(lastError);
        this.metrics.errorClassCounts[errorClass] =
          (this.metrics.errorClassCounts[errorClass] || 0) + 1;

        // Get strategy for error class
        strategy = this.getStrategy(errorClass);

        // Check if should retry
        if (!strategy.shouldRetry) {
          this.metrics.failedAfterRetries++;
          throw new Error(`Non-retryable error (${errorClass}): ${lastError.message}`);
        }

        // Check if exceeded max attempts for this error class
        if (attempt >= strategy.maxAttempts) {
          break;
        }

        // Track retry
        this.metrics.totalRetries++;
        this.budget.retryCount++;

        // Calculate backoff with jitter
        let baseBackoff = strategy.initialBackoff * Math.pow(strategy.backoffMultiplier, attempt - 1);
        if (baseBackoff > strategy.maxBackoff) {
          baseBackoff = strategy.maxBackoff;
        }
        const backoff = this.calculateBackoff(baseBackoff, attempt);

        // Sleep with backoff
        await new Promise((resolve) => setTimeout(resolve, backoff));
      }
    }

    // All attempts failed
    this.metrics.failedAfterRetries++;
    throw new Error(
      `Max retry attempts (${strategy.maxAttempts}) exceeded for ${errorClass}: ${lastError?.message}`
    );
  }

  /**
   * Get current metrics.
   */
  getMetrics(): EnhancedRetryMetrics {
    return this.metrics;
  }
}

File: agenkit-rust/src/infrastructure/retry_enhanced.rs
// Enhanced retry logic with:
// - Multiple jitter types (Full, Equal, Decorrelated)
// - Per-error-type retry strategies
// - Budget awareness (cost and count limits)
// - Backpressure detection
// - Detailed metrics

use crate::core::{Agent, AgentError, Message};
use async_trait::async_trait;
use rand::Rng;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;

/// Jitter types for retry backoff.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum JitterType {
    None,
    Full,
    Equal,
    Decorrelated,
}

/// Error classification for retry strategies.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ErrorClass {
    Transient,
    RateLimit,
    Timeout,
    ServerError,
    ClientError,
    Unknown,
}

impl ErrorClass {
    pub fn as_str(&self) -> &str {
        match self {
            ErrorClass::Transient => "transient",
            ErrorClass::RateLimit => "rate_limit",
            ErrorClass::Timeout => "timeout",
            ErrorClass::ServerError => "server_error",
            ErrorClass::ClientError => "client_error",
            ErrorClass::Unknown => "unknown",
        }
    }
}

/// Retry strategy for specific error class.
#[derive(Debug, Clone)]
pub struct ErrorStrategy {
    pub error_class: ErrorClass,
    pub max_attempts: usize,
    pub initial_backoff: Duration,
    pub max_backoff: Duration,
    pub backoff_multiplier: f64,
    pub should_retry: bool,
}

/// Retry budget to limit costs.
#[derive(Debug, Clone)]
pub struct RetryBudget {
    pub max_cost: f64,
    pub current_cost: f64,
    pub max_retries_per_hour: u64,
    pub retry_count: u64,
    pub window_start: Instant,
}

/// Enhanced retry configuration.
#[derive(Debug, Clone)]
pub struct EnhancedRetryConfig {
    // Basic retry settings
    pub max_attempts: usize,
    pub initial_backoff: Duration,
    pub max_backoff: Duration,
    pub backoff_multiplier: f64,

    // Jitter settings
    pub jitter_type: JitterType,
    pub jitter_min_ratio: f64,

    // Error-specific strategies
    pub error_strategies: HashMap<ErrorClass, ErrorStrategy>,

    // Budget settings
    pub enable_budget: bool,
    pub max_cost_per_hour: f64,
    pub max_retries_per_hour: u64,

    // Backpressure detection
    pub enable_backpressure: bool,
    pub backpressure_threshold: f64,
    pub backpressure_window: usize,
}

impl Default for EnhancedRetryConfig {
    fn default() -> Self {
        let mut error_strategies = HashMap::new();

        error_strategies.insert(
            ErrorClass::Transient,
            ErrorStrategy {
                error_class: ErrorClass::Transient,
                max_attempts: 5,
                initial_backoff: Duration::from_millis(100),
                max_backoff: Duration::from_secs(5),
                backoff_multiplier: 2.0,
                should_retry: true,
            },
        );

        error_strategies.insert(
            ErrorClass::RateLimit,
            ErrorStrategy {
                error_class: ErrorClass::RateLimit,
                max_attempts: 10,
                initial_backoff: Duration::from_secs(60),
                max_backoff: Duration::from_secs(300),
                backoff_multiplier: 1.5,
                should_retry: true,
            },
        );

        error_strategies.insert(
            ErrorClass::Timeout,
            ErrorStrategy {
                error_class: ErrorClass::Timeout,
                max_attempts: 3,
                initial_backoff: Duration::from_secs(2),
                max_backoff: Duration::from_secs(30),
                backoff_multiplier: 2.0,
                should_retry: true,
            },
        );

        error_strategies.insert(
            ErrorClass::ServerError,
            ErrorStrategy {
                error_class: ErrorClass::ServerError,
                max_attempts: 3,
                initial_backoff: Duration::from_secs(5),
                max_backoff: Duration::from_secs(60),
                backoff_multiplier: 2.0,
                should_retry: true,
            },
        );

        error_strategies.insert(
            ErrorClass::ClientError,
            ErrorStrategy {
                error_class: ErrorClass::ClientError,
                max_attempts: 1,
                initial_backoff: Duration::from_secs(0),
                max_backoff: Duration::from_secs(0),
                backoff_multiplier: 1.0,
                should_retry: false,
            },
        );

        Self {
            max_attempts: 3,
            initial_backoff: Duration::from_secs(1),
            max_backoff: Duration::from_secs(30),
            backoff_multiplier: 2.0,
            jitter_type: JitterType::Full,
            jitter_min_ratio: 0.5,
            error_strategies,
            enable_budget: false,
            max_cost_per_hour: 100.0,
            max_retries_per_hour: 1000,
            enable_backpressure: true,
            backpressure_threshold: 0.5,
            backpressure_window: 100,
        }
    }
}

/// Enhanced retry metrics.
#[derive(Debug, Clone, Default)]
pub struct EnhancedRetryMetrics {
    pub total_attempts: u64,
    pub successful_first_attempt: u64,
    pub successful_on_retry: u64,
    pub failed_after_retries: u64,
    pub total_retries: u64,
    pub total_jitter_added: f64,
    pub budget_exceeded_count: u64,
    pub backpressure_detected: u64,
    pub error_class_counts: HashMap<ErrorClass, u64>,
    pub recent_results: Vec<bool>,
}

/// Enhanced retry decorator wraps an agent with enhanced retry logic.
pub struct EnhancedRetryDecorator {
    agent: Arc<dyn Agent>,
    config: EnhancedRetryConfig,
    metrics: Arc<RwLock<EnhancedRetryMetrics>>,
    budget: Arc<RwLock<RetryBudget>>,
}

impl EnhancedRetryDecorator {
    /// Create a new enhanced retry decorator.
    pub fn new(agent: Arc<dyn Agent>, config: EnhancedRetryConfig) -> Self {
        let budget = RetryBudget {
            max_cost: config.max_cost_per_hour,
            current_cost: 0.0,
            max_retries_per_hour: config.max_retries_per_hour,
            retry_count: 0,
            window_start: Instant::now(),
        };

        Self {
            agent,
            config,
            metrics: Arc::new(RwLock::new(EnhancedRetryMetrics::default())),
            budget: Arc::new(RwLock::new(budget)),
        }
    }

    fn classify_error(&self, error: &AgentError) -> ErrorClass {
        let err_str = error.to_string().to_lowercase();

        if err_str.contains("rate limit") || err_str.contains("429") {
            ErrorClass::RateLimit
        } else if err_str.contains("timeout") || err_str.contains("timed out") {
            ErrorClass::Timeout
        } else if err_str.contains("500") || err_str.contains("502") || err_str.contains("503") {
            ErrorClass::ServerError
        } else if err_str.contains("400")
            || err_str.contains("401")
            || err_str.contains("403")
            || err_str.contains("404")
        {
            ErrorClass::ClientError
        } else {
            ErrorClass::Unknown
        }
    }

    fn get_strategy(&self, error_class: ErrorClass) -> ErrorStrategy {
        self.config
            .error_strategies
            .get(&error_class)
            .cloned()
            .unwrap_or_else(|| ErrorStrategy {
                error_class,
                max_attempts: self.config.max_attempts,
                initial_backoff: self.config.initial_backoff,
                max_backoff: self.config.max_backoff,
                backoff_multiplier: self.config.backoff_multiplier,
                should_retry: true,
            })
    }

    fn calculate_backoff(&self, base_backoff: Duration, attempt: usize) -> Duration {
        let base_ms = base_backoff.as_millis() as f64;
        let mut rng = rand::thread_rng();

        let jittered_ms = match self.config.jitter_type {
            JitterType::None => base_ms,
            JitterType::Full => {
                let jittered = rng.gen::<f64>() * base_ms;
                jittered
            }
            JitterType::Equal => {
                let min_backoff = base_ms * self.config.jitter_min_ratio;
                min_backoff + rng.gen::<f64>() * (base_ms - min_backoff)
            }
            JitterType::Decorrelated => {
                if attempt == 1 {
                    base_ms
                } else {
                    let previous = self.calculate_backoff(base_backoff, attempt - 1);
                    let previous_ms = previous.as_millis() as f64;
                    let jittered = rng.gen::<f64>() * previous_ms * 3.0 + base_ms;
                    let max_ms = self.config.max_backoff.as_millis() as f64;
                    if jittered > max_ms {
                        max_ms
                    } else {
                        jittered
                    }
                }
            }
        };

        Duration::from_millis(jittered_ms as u64)
    }

    async fn check_budget(&self, _cost: f64) -> bool {
        if !self.config.enable_budget {
            return true;
        }

        let mut budget = self.budget.write().await;

        // Reset window if hour has passed
        if budget.window_start.elapsed() > Duration::from_secs(3600) {
            budget.current_cost = 0.0;
            budget.retry_count = 0;
            budget.window_start = Instant::now();
        }

        // Check cost budget
        if budget.current_cost + _cost > budget.max_cost {
            let mut metrics = self.metrics.write().await;
            metrics.budget_exceeded_count += 1;
            return false;
        }

        // Check retry count budget
        if budget.retry_count >= budget.max_retries_per_hour {
            let mut metrics = self.metrics.write().await;
            metrics.budget_exceeded_count += 1;
            return false;
        }

        true
    }

    async fn check_backpressure(&self) -> bool {
        if !self.config.enable_backpressure {
            return false;
        }

        let metrics = self.metrics.read().await;
        if metrics.recent_results.len() < self.config.backpressure_window {
            return false;
        }

        // Calculate failure rate
        let failures = metrics.recent_results.iter().filter(|&&success| !success).count();
        let failure_rate = failures as f64 / metrics.recent_results.len() as f64;

        failure_rate > self.config.backpressure_threshold
    }

    /// Get current metrics.
    pub async fn get_metrics(&self) -> EnhancedRetryMetrics {
        self.metrics.read().await.clone()
    }
}

#[async_trait]
impl Agent for EnhancedRetryDecorator {
    fn name(&self) -> String {
        self.agent.name()
    }

    fn capabilities(&self) -> Vec<String> {
        self.agent.capabilities()
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        let mut last_error: Option<AgentError> = None;
        let mut error_class = ErrorClass::Unknown;
        let mut strategy = self.get_strategy(error_class);

        for attempt in 1..=self.config.max_attempts {
            {
                let mut metrics = self.metrics.write().await;
                metrics.total_attempts += 1;
            }

            // Check budget before attempt
            if self.config.enable_budget {
                if !self.check_budget(0.0).await {
                    return Err(AgentError::ExecutionError(
                        "Retry budget exceeded".to_string(),
                    ));
                }
            }

            // Check backpressure
            if self.check_backpressure().await {
                tokio::time::sleep(Duration::from_secs(5)).await;
            }

            // Process message
            match self.agent.process(message.clone()).await {
                Ok(response) => {
                    // Success
                    let mut metrics = self.metrics.write().await;
                    if attempt == 1 {
                        metrics.successful_first_attempt += 1;
                    } else {
                        metrics.successful_on_retry += 1;
                    }
                    metrics.recent_results.push(true);
                    if metrics.recent_results.len() > self.config.backpressure_window {
                        metrics.recent_results.remove(0);
                    }

                    return Ok(response);
                }
                Err(error) => {
                    // Failure
                    last_error = Some(error.clone());

                    // Track failure for backpressure
                    {
                        let mut metrics = self.metrics.write().await;
                        metrics.recent_results.push(false);
                        if metrics.recent_results.len() > self.config.backpressure_window {
                            metrics.recent_results.remove(0);
                        }
                    }

                    // Classify error
                    error_class = self.classify_error(&error);
                    {
                        let mut metrics = self.metrics.write().await;
                        *metrics.error_class_counts.entry(error_class).or_insert(0) += 1;
                    }

                    // Get strategy for error class
                    strategy = self.get_strategy(error_class);

                    // Check if should retry
                    if !strategy.should_retry {
                        let mut metrics = self.metrics.write().await;
                        metrics.failed_after_retries += 1;
                        return Err(AgentError::ExecutionError(format!(
                            "Non-retryable error ({}): {}",
                            error_class.as_str(),
                            error
                        )));
                    }

                    // Check if exceeded max attempts for this error class
                    if attempt >= strategy.max_attempts {
                        break;
                    }

                    // Track retry
                    {
                        let mut metrics = self.metrics.write().await;
                        metrics.total_retries += 1;
                    }

                    {
                        let mut budget = self.budget.write().await;
                        budget.retry_count += 1;
                    }

                    // Calculate backoff with jitter
                    let base_backoff_ms =
                        strategy.initial_backoff.as_millis() as f64 * strategy.backoff_multiplier.powi(attempt as i32 - 1);
                    let mut base_backoff = Duration::from_millis(base_backoff_ms as u64);
                    if base_backoff > strategy.max_backoff {
                        base_backoff = strategy.max_backoff;
                    }
                    let backoff = self.calculate_backoff(base_backoff, attempt);

                    // Sleep with backoff
                    tokio::time::sleep(backoff).await;
                }
            }
        }

        // All attempts failed
        {
            let mut metrics = self.metrics.write().await;
            metrics.failed_after_retries += 1;
        }

        Err(AgentError::ExecutionError(format!(
            "Max retry attempts ({}) exceeded for {}: {}",
            strategy.max_attempts,
            error_class.as_str(),
            last_error.unwrap_or_else(|| AgentError::ExecutionError("Unknown error".to_string()))
        )))
    }
}
